### ISSUE TITLE
Make mode constructors/destructors composable

### ISSUE THREAD

**Issue by aartaka** [atlas-engineer/nyxt#1005]
One may need to run the constructors of parent modes in while initializing child classes. 

For example: 
It is a good practice in Nyxt to create small modes for settings-toggling, like `noimage-mode` and `noscript-mode`. I'm thinking about `noclipboard-mode` to disable clipboard access for JS. This should be a separate mode. But it should be part of `reduce-tracking-mode` too, because clipboard access is a way of tracking and breaching one's privacy and security.

noclipboard-mode.lisp (not an actual code, no renderer abstraction):
``` lisp
(define-mode noclipboard-mode ()
  "Disable JavaScript access to clipboard."
  ((destructor
    (lambda (mode)
      (setf (webkit:webkit-settings-javascript-can-access-clipboard
             (webkit:webkit-web-view-get-settings (gtk-object buffer)))
            t)))
   (constructor
    (lambda (mode)
      (setf (webkit:webkit-settings-javascript-can-access-clipboard
             (webkit:webkit-web-view-get-settings (gtk-object buffer)))
            nil)))))
```

reduce-tracking-mode.lisp:
``` lisp
(define-mode reduce-tracking-mode (noclipboard-mode root-mode)
  "Set specific settings in the web view in order to mitigate fingerprinting,
(how third-party trackers attempt to indentify you."
  ((preferred-languages '("en_US")
                        :type list-of-strings
                        :documentation "The list of languages that will be sent as
part of the Accept-Language HTTP header.")
   (destructor
   ;;; noclipboard-mode's destructor should run too, this code shouldn't be there
   ;; (setf (webkit:webkit-settings-javascript-can-access-clipboard
   ;;        (webkit:webkit-web-view-get-settings (gtk-object buffer)))
   ;;       t)))
    (lambda (mode)
      (ffi-set-preferred-languages (buffer mode)
                                   (list (first
                                          (str:split
                                           "."
                                           (or (uiop:getenv "LANG") "")))))))
   (constructor
   ;;; noclipboard-mode's constructor should run too, this code shouldn't be there
   ;; (setf (webkit:webkit-settings-javascript-can-access-clipboard
   ;;        (webkit:webkit-web-view-get-settings (gtk-object buffer)))
   ;;       nil)))
    (lambda (mode)
      (ffi-set-preferred-languages (buffer mode)
                                   (preferred-languages mode))))))
```

So, we need to run the same clipboard-disabling code in two mode constructors. It'll be nice to have `constuctor`/`destructor` inherited and composed (`progn` method combination) to not duplicate code. It also seems to be relatively intuitive (if it's not, then, please, soberize me). 

There's no explicit way to do so in conventional CL, as far as I know, so adding this feature would require hooking into `define-class` yet again. The benefit of actually having modes inherit settings and side-effects of each other seems to be more than the expences :)

How does that sound to you?

**Comment by Ambrevar** [comment]
Thanks for bringing this up.

Actually this comment has been in mode.lisp since the beginning:

```
; TODO: Make constructor / destructor methods?  Then we can use initialize-instance, etc.
```

Then in the constructor we can call (call-next-method) to use the parent constructor.

Another thought: we could replace the constructor by
`initialize-instance :after`.
The drawback is that it's a bit more cumbersome to write but most
importantly, there is no equivalent for the destructor (correct me if
I'm wrong).

**Comment by Ambrevar** [comment]
I suggest this:

- Leave `constructor` and `destructor` where they are.

- Define a `construct` method that's called when instantiating the mode.
- The default `construct` method runs `(maybe-funcall (constructor mode) mode)`.

- Replaces calls to `constructor` or `destructor` in the `define-mode` mode macro (i.e. the mode command) by `construct` and `destroy` respectively.

The above allows us to use function dispatch for mode construction / destruction, while also allowing for customization via slots, which is useful for small modes that are not going to be inherited from.

Thoughts?
@jmercouris @aartaka 

If you agree, I'll implement the same strategy for the prompt buffer in #1157.

**Comment by aartaka** [comment]
That seems all right to me :)

**Comment by jmercouris** [comment]
Hm, seems like it makes some sense. I almost wish there was a 'destroy-instance' complement to 'initialize-instance', because it very well feels like we are reinventing the wheel here :-D

**Comment by jmercouris** [comment]
I can't think of anything better, so let's take this approach for now :-)!

**Comment by Ambrevar** [comment]
Actually...

What if we _didn't_ add `construct`, but used `initialize-instance
:after` instead?  

Then we add `destroy-instance` ourselves, which calls the destroy method
by default.  The only syntactical oddity is that `destroy-instance` does
not take the `:after`, but I guess that's OK.

**Comment by jmercouris** [comment]
That does sound even better. Yes!

**Comment by jmercouris** [comment]
Maybe with function advising it *can* take an after?

**Comment by jmercouris** [comment]
Actually what would stop us from writing an after for a destroy-instance method? I don't see why that wouldn't work.

**Comment by jmercouris** [comment]
We don't need function advising since it is a method and we can use :after!

**Comment by Ambrevar** [comment]
Do we actually need a `constructor` and `destructor`?  Originally, we had them for the following reasons:

1. Because there is no destruction method in CLOS (but it's easy to add one).
2. Because then the define-mode macro is trivial to write to customize the constructor / destructor.
3. Because I wanted to customize the construction / destruction of mode instances on the fly.

This raises a few issues:

- We don't have `constructor` / `destructor` dispatch, as hinted by the original post.
- We duplicate CLOS but poorly.
- It's harder to understand.

Maybe an answer to it all is to not use these slots at all.

1. We can define `destroy-instance` on `root-mode` which calls `(call-next-method)`.  Voilà!
2. We can update `define-macro` to support new class options: `initializer` and `destroyer` which would set the `initialize-instance` and `destroy-instance` methods, respectively.
3. We actually never customize mode instances on the fly.  Should we ever need to do it, the solution is easy: make the instance then edit the returned object in the calling code.  If we don't like what the initializer runs, then we should probably define a new mode anyways.

Thoughts?

**Comment by Ambrevar** [comment]
@aadcg You might be interested by my last proposal ;)

**Comment by aadcg** [comment]
> - We can define destroy-instance on root-mode which calls
>   (call-next-method). Voilà!

Smart.

> - We can update define-macro to support new class options: initializer
>   and destroyer which would set the initialize-instance and
>   destroy-instance methods, respectively.

You meant define-class right?  I think I followed you.

> - We actually never customize mode instances on the fly. Should we ever
>   need to do it, the solution is easy: make the instance then edit the
>   returned object in the calling code. If we don't like what the
>   initializer runs, then we should probably define a new mode anyways.

Diff-mode could use customising instances on the fly.  Say I have an
instance of diff-mode, which has a diff-html slot that keeps the html as
str.  If I want to show the diff not "line-wise" but "column-wise", I'd
have to change the slot of that instance.

@jmercouris what are your thoughts?

**Comment by Ambrevar** [comment]
> You meant define-class right? I think I followed you.

I meant `define-mode`.

For diff-mode, no problem, you just instantiate it with the right option like any other class:

```lisp
(make-instance 'diff-mode :display-mode :line-wise)
```

**Comment by jmercouris** [comment]
I like Pierre's suggestions personally :-)

**Comment by BlueFlo0d** [comment]
> I suggest this:
> 
> * Leave `constructor` and `destructor` where they are.
> * Define a `construct` method that's called when instantiating the mode.
> * The default `construct` method runs `(maybe-funcall (constructor mode) mode)`.
> * Replaces calls to `constructor` or `destructor` in the `define-mode` mode macro (i.e. the mode command) by `construct` and `destroy` respectively.
> 
> The above allows us to use function dispatch for mode construction / destruction, while also allowing for customization via slots, which is useful for small modes that are not going to be inherited from.

I don't think we need to retain the slot to allow customization, it sounds rather ad-hoc. If we're going to use CLOS generic function, then the user can just customize by defining methods specialized on `*-user-class`. No need to add two slots to every mode

I also like `construct` and `destruct` (or, `on-enable` and `on-disable`) better than `initialize-instance` and `destroy-instance`. It sounds a bit awkward to add a "counterpart" to a built-in generic function, almost like we're extending the CL standard. And I think creating a mode object should be decoupled from attaching to a buffer and enabling it. I can imagine some magic that create a mode first using `make-instance`, then explicitly manipulating it in some way (e.g. changing some slots), and finally enabling it. Separating `initialize-instance` with `construct/on-enable` makes this possible.

**Comment by aartaka** [comment]
How about hooking into `trivial-gc:finalize`? It's portable (although I didn't find any trace of CCL-specific code in it), intuitive enough and works at object garbage collection, which is synonymous to destruction.

**Comment by BlueFlo0d** [comment]
> How about hooking into `trivial-gc:finalize`? It's portable (although I didn't find any trace of CCL-specific code in it), intuitive enough and works at object garbage collection, which is synonymous to destruction.

I think exactly the opposite, and your point actually confirms this -- mode enabling/disabling is not synonymous to object construction/destruction.

Suppose we treat them as synonymous, then it's very logical to do the above -- put mode disabling code inside `trivial-gc:finalize`. However this is rarely the intended semantics for a modal user interface. Code inside `finalize` will run at unpredictable time (during garbage collection) rather than right after user executes some mode toggling command.

**Comment by BlueFlo0d** [comment]
After some second thoughts, I think maybe we can even go further by providing 4 (instead of 2!) generic functions: `mode-enable`, `mode-disable`, `mode-pause` and `mode-resume`. `mode-enable/mode-disable` will always use freshly initialized mode objects so that they're useful to reset all states of mode in case of malfunction, and `mode-pause/resume` will instead reuse existing mode object, so one can temporarily disable a mode without losing its states. This is fully optional though, and maybe there're some more sensible choices for the function names.

**Comment by Ambrevar** [comment]
Indeed, 4 methods would give us maximal customizability.

@aartaka: We can't use tg:finalize because we cannot access the object
in that method.

**Comment by aartaka** [comment]
@Ambrevar, we can, but we need to change the signature of `destructor` to have no arguments. In some modes it will be trivial, in some — not so.

**Comment by Ambrevar** [comment]
I'm not sure it's well defined to handle values stored in the modes
beying destroyed.

**Comment by BlueFlo0d** [comment]
> Indeed, 4 methods would give us maximal customizability. @aartaka: We can't use tg:finalize because we cannot access the object in that method.

It's a bit tricky to hide the additional complexity when it's not needed though, both for user and the developer. 

What I currently have in mind: we just provide a two `enable/disable` generic functions, which is suppose to correctly enable/disable an existing mode object. Any work that should be done to initialize a fresh new mode object should be in `initialize-instance`. The mode toggler now accept one optional `C-u` prefix, which functions as a "reset flag": if prefix is not present, it always try to reuse existing mode object. If it is presented, it always allocate a fresh mode object when enabling, and discard mode object when disabling.

**Comment by Ambrevar** [comment]
I'll work on it next and go forward with @BlueFlo0d last proposal.
The `C-u` prefix can be implemented as a second step.

**Comment by Ambrevar** [comment]
Fixed with #2200.  See #2209 for a follow up discussion on mode destruction.
